### PR TITLE
Hubs Frontend

### PR DESCRIPTION
**Pull request by barrioskevin**
*Please complete the checklist so the group can quickly understand your PR.*

### Category
- [X] **Feature**
- [ ] **Bug Fix**
- [ ] **Testing**
- [ ] **Documentation**
- [ ] **Other**

### Domain
- [X] **Frontend**
- [ ] **Backend**
- [ ] **Both**

### Description
This pull request is my attempt to introduce some type of front end work for hubs. I make new pages where we can see all the hubs in rebel hub and also create new hubs. All pages are protected under protected route. 

### Related Issue(s)
Although not the home page i am targeting issue #50 since i believe the hubs page might be able to act like a homepage. But if not we can leave this issue open. I didn't find any issue regarding a hubs page.

### New Additions
I added 4 new components along with their css under 
`frontend/src/components/hubs/*` 
I use these components when building the hub page. They are mostly specific to hubs but the two side view component has potential for use cases beyond hubs.

I also added 2 new pages under
`frontend/src/app/hubs/*`
One is the main hubs page `page.js` where you can browse the different hubs on rebel hub. There is also a create hub section on this page.

The next page i added was the hubs [id] page where it will show the hub page for a specific id. This page still looks kinda ugly in terms of design but it includes all the things needed when looking at a specific hub. we should probably expand more on this page including in more of our apps components currently i only included the posts lists component for displaying the hub's posts.

### Modifications
I modified `backend/hubs/serializers.py` in order for the create hub response to return an id. I use the id to then redirect to the new created hub page. 

I modified `backend/users/serializer.py` ,  `backend/users/views.py` , `backend/users/urls.py` to add in a simple public info endpoint so we can get the usernames of different users based on an id.

I also modified the navbar component to show a display name, this will be the user's username.

### New Dependencies
No new dependencies.

### Visuals
image
image
image

### Architecture of Your Changes
I decided to make the posts lists component be able to render all the hubs in the limited view we only show the data thats returned when calling `api/hubs/` and then if a user clicks on a hub they will be redirected to the individual hubs page where they can see the hub in more detail. 

The hub page also includes a section where a user can create their own hub.

For a detailed hub page I added a few things such as the posts lists component showing all the hubs posts. and a members lists where we can see all the members. Currently the members lists will only show your name since there is no endpoint for getting detailed info about a specific user.

There is a background image on the hubs detail page where it can be changed in the future to be the hubs image. 

### Usage and Integration
goto the main site `/hubs/` to see the hubs page. goto `hubs/id` where id is an id of a hub, to the hubs detailed page.

### Testing
I made sure all tests passed before opening PR. The rest of the frontend testing was done through trial and error.

### Mentions
Nobody specific to mention but I would appreciate any feedback on the hubs page design and user experience.

### CODE REVIEW

**Review by leed85 — APPROVED** [review]

**Comment by misherwa** [comment]
Hey Kevin, I think it was a good call to change the serializers.py for the hub response to return an id instead of the name since there could be hubs with the same name.  

Because of that, I was wondering if it's possible to consider a feature that makes sure no two hubs share the same name. If you want to consider this, I can try helping you and it seems that you would have to update your models.py and serializers.py. But, it would allow no two hubs with the same name to exist. If two hubs were named the same, then there isn't much to differentiate the two hubs unless a user goes inside and checks it. Also, it seems like this is what Reddit does, where you can't have two or more subreddits with the same name. For example, there is only one "r/college".

I was also wondering how you set up your page. I tried running the backend and frontend servers. I also created a super user, added some hubs and posts, and used "/admin" in the backend. However, I was having trouble with the frontend, where I could go into the login page using: "/users/login" but I ended up with a network error if I login. If I try putting "/hubs" it sends me to the login page.

**Comment by barrioskevin** [comment]
you already can't make hubs with the same name.

The hubs page is a protected route so if you are not signed it you cannot access it and it will redirect you to the login page.

hubs page `http://localhost:3000/hubs/`

**Comment by misherwa** [comment]
> you already can't make hubs with the same name.


Sorry, you're right about not making hubs with the same name. I checked the backend and this is the result:
w

But for the frontend, it doesn't give a warning message like how the backend does. For example, I created a hub named CS 135 and tried hitting the "Create Hub!" button. But since there already was a CS 135 hub, it didn't get added:
q2

I'd recommend adding a similar warning message to how the backend does. Just in case a user doesn't know or can't see that a Hub already has the same name as the one they want to create.

Also, I was double-checking in the backend if anything else can have the same name, and it looks like only the title for posts can have the same name.

Here's an example:
cur

Above, you'll see that there are 4 different Posts with the same title "classroom". The last title has a capital "C" at the start, but they all belong in the same Hub. I also tested this with a different author by creating another user, and it still lets me have the same Post title. Would you consider this an issue? If so, then the same name restriction to Posts can be applied by either this PR, or we can make a new issue and apply it from a different PR.

Also, this is a view of the frontend:
cur

File: frontend/src/components/posts/buttons/create-post-button.tsx [hunk 1]
```diff
@@ -3,6 +3,10 @@
 import Link from "next/link";
 import styles from "./post-buttons.module.css"
 import { gotoCreatePostPage } from "@/utils/url-segments";
+=======
+import styles from "./post-buttons.module.css"
+import { gotoCreatePostPage } from "@/utils/posts/url-segments";
+>>>>>>> b905c44 (Update create-post-button.tsx)
```
Comment (thecatfour) [inline frontend/src/components/posts/buttons/create-post-button.tsx]: I think you forgot to resolve the merge conflicts in create-post-button.tsx. The new lines
```
=======
import styles from "./post-buttons.module.css"
import { gotoCreatePostPage } from "@/utils/posts/url-segments";
>>>>>>> b905c44 (Update create-post-button.tsx)
```
should be deleted.

**Comment by thecatfour** [comment]
To fix the edit hub input bug, you need to change the inputs from:
image
To:
image
Basically, I just moved the input components out of the functions and into the main return component.

**Comment by thecatfour** [comment]
Is anyone else having an issue with adding a member to a hub? Here is a gif of me showing how every other feature of being a hub owner works. However, when I try to add a member using the ADD button, I get a HTTP 400 error. 

hubs issue

Is anyone else getting this error, or is it just me?

For context, I used the admin panel to add those members to the pending list.

**Comment by thecatfour** [comment]
Compared to the other things I found, this issue isn't as important. When I open up the browse hubs page on a somewhat smaller screen, the hub description input goes off the component.

image

It is fine when the browser takes up my full screen, but I think it is a good idea to bring attention to this. Should we just leave it for now?

**Comment by thecatfour** [comment]
> Above, you'll see that there are 4 different Posts with the same title "classroom". The last title has a capital "C" at the start, but they all belong in the same Hub. I also tested this with a different author by creating another user, and it still lets me have the same Post title. Would you consider this an issue? If so, then the same name restriction to Posts can be applied by either this PR, or we can make a new issue and apply it from a different PR.

YouTube and Reddit both allow videos/posts to have the same title, so I do not consider this an issue.

**Comment by misherwa** [comment]
> YouTube and Reddit both allow videos/posts to have the same title, so I do not consider this an issue.

yeah, looks like you're right about the same title for videos/posts being allowed.

My other recommendation is adding an error message in the frontend when a user tries to create a hub with the same name as another hub

**Comment by barrioskevin** [comment]
@misherwa @thecatfour 
I believe I have addressed all of your issues. Please let me know if there's anything else I should fix.
Thank you,

**Review by misherwa — APPROVED** [review]

**Comment by thecatfour** [comment]
Hub moderators still have the button that lets them kick members, but it just gives a 400 error when clicked. I think you should either hide the button from moderators or allow them to kick people.

**Comment by nnguyen144** [comment]
@thecatfour 's issue of hub moderating seeing the delete member button but not being able to has been fixed so that hub moderators cannot see the button. The PR can thus now be merged.

**Pull request merged**